File: lib/network/packets/server/s_property_pack.dart
/// 單一場景物件（來自伺服器 property 表）。
///
/// [bubbleText] 不隨本封包預載 —— 互動時才由 S_BUBBLE_DIALOG 單獨送出。
class PropertyObject {
  const PropertyObject({
    required this.objId,
    required this.name,
    required this.x,
    required this.y,
    required this.heading,
    required this.propertyId,
    required this.pngid,
    required this.blocking,
    required this.footprintW,
    required this.footprintH,
    required this.offsetX,
    required this.offsetY,
    required this.action,
    required this.actionType,
    required this.value,
  });

  /// 世界唯一物件編號（狀態更新／移除封包都以此對應）。
  final int objId;
  final String name;
  final int x;
  final int y;
  final int heading;

  /// property 表的模板編號。
  final int propertyId;

  /// 圖片編號，對應 assets/data/object_catalog.json 的物件 id。
  final int pngid;

  /// 是否阻擋通行（伺服器權威）。
  final bool blocking;

  /// 碰撞佔格寬（地面格數，非視覺尺寸）。
  /// 視覺大小由 object_catalog 的圖與 tilesW 決定，兩者不可混用。
  final int footprintW;

  /// 碰撞佔格高（地面格數，非視覺尺寸）。
  final int footprintH;

  /// 像素微調（玩家家具用；壁掛物以負的 offsetY 往上推到牆面高度）。
  final int offsetX;
  final int offsetY;

  /// 是否可互動（採空後伺服器會改為 false）。
  final bool action;

  /// 互動方式：1=對話、2=採集；0=不可互動。
  final int actionType;

  /// 進度條之類的數值。
  final int value;

  PropertyObject copyWith({int? value, bool? action}) => PropertyObject(
        objId: objId,
        name: name,
        x: x,
        y: y,
        heading: heading,
        propertyId: propertyId,
        pngid: pngid,
        blocking: blocking,
        footprintW: footprintW,
        footprintH: footprintH,
        offsetX: offsetX,
        offsetY: offsetY,
        action: action ?? this.action,
        actionType: actionType,
        value: value ?? this.value,
      );

  factory PropertyObject.fromJson(Map<String, dynamic> j) => PropertyObject(
        objId: (j['objId'] as num?)?.toInt() ?? 0,
        name: j['name'] as String? ?? '',
        x: (j['x'] as num?)?.toInt() ?? 0,
        y: (j['y'] as num?)?.toInt() ?? 0,
        heading: (j['heading'] as num?)?.toInt() ?? 2,
        propertyId: (j['propertyId'] as num?)?.toInt() ?? 0,
        pngid: (j['pngid'] as num?)?.toInt() ?? 0,
        blocking: j['blocking'] as bool? ?? false,
        footprintW: (j['footprintW'] as num?)?.toInt() ?? 1,
        footprintH: (j['footprintH'] as num?)?.toInt() ?? 1,
        offsetX: (j['offsetX'] as num?)?.toInt() ?? 0,
        offsetY: (j['offsetY'] as num?)?.toInt() ?? 0,
        action: j['action'] as bool? ?? false,
        actionType: (j['actionType'] as num?)?.toInt() ?? 0,
        value: (j['value'] as num?)?.toInt() ?? 0,
      );
}

/// 地圖上的場景物件清單。
class SPropertyPack {
  const SPropertyPack({required this.mapId, required this.properties});

  final int mapId;
  final List<PropertyObject> properties;

  factory SPropertyPack.fromData(Map<String, dynamic> data) {
    final raw = data['properties'] as List<dynamic>? ?? const [];
    return SPropertyPack(
      mapId: (data['mapId'] as num?)?.toInt() ?? 0,
      properties: raw
          .map((p) => PropertyObject.fromJson(p as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: src/main/java/com/xin/server/packet/server/S_PropertyPack.java
package com.xin.server.packet.server;

import java.util.List;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.xin.server.model.instance.DecorationInstance;
import com.xin.server.model.instance.PropertyInstance;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;
import com.xin.server.world.WorldProperty;

/**
 * 場景物件封包（對應 opcode {@code S_PROPERTY_PACK}）。
 * <p>
 * 資料來自 {@code property} 表，經 {@code spawnlist_scene}
 * 的設定生成。{@code actionType}：1=對話、2=採集。
 * <p>
 * {@code footprintW/H} 是<b>地面佔格</b>（碰撞用），不是視覺尺寸；
 * {@code pngid} 對應前端 {@code object_catalog.json} 的物件 id。
 * <p>
 * 注意：{@code bubbleText} <b>不</b>包含在此封包中——整張圖的對話文字沒必要預載，
 * 實際互動時才由 {@link S_BubbleDialog} 單獨送出。
 *
 * <pre>
 * {
 *   "op": "S_PROPERTY_PACK",
 *   "data": {
 *     "mapId": 1,
 *     "properties": [
 *       { "objId": 2000000003, "name": "石碑", "x": 15, "y": 18, "heading": 2,
 *         "propertyId": 1001, "pngid": 1010, "blocking": true,
 *         "footprintW": 1, "footprintH": 1,
 *         "action": true, "actionType": 1, "value": 0 }
 *     ]
 *   }
 * }
 * </pre>
 */
public class S_PropertyPack extends ServerBasePacket {

    private S_PropertyPack(int mapId) {
        super(ServerOpcodes.S_PROPERTY_PACK);
        put("mapId", mapId);

        List<PropertyInstance> properties = WorldProperty.get().getPropertiesByMap(mapId);
        ArrayNode arr = newArray();
        for (PropertyInstance property : properties) {
            ObjectNode node = newObject();
            node.put("objId",      property.getId());
            node.put("name",       property.getName());
            node.put("x",          property.getX());
            node.put("y",          property.getY());
            node.put("heading",    property.getHeading());
            node.put("propertyId", property.getPropertyTemplateId());
            node.put("pngid",      property.getPngId());
            node.put("blocking",   property.isBlocking());
            node.put("footprintW", property.getFootprintW());
            node.put("footprintH", property.getFootprintH());
            node.put("offsetX",    0);
            node.put("offsetY",    0);
            node.put("action",     property.isAction());
            node.put("actionType", property.getActionType());
            node.put("value",      property.getValue());
            arr.add(node);
        }
        putArray("properties", arr);
    }

    /**
     * 玩家裝飾版本：欄位與世界物件完全相同，前端因此只需要一條解析路徑，
     * 不必分辨這件東西是企劃擺的還是玩家自己擺的。
     */
    private S_PropertyPack(int mapId, List<DecorationInstance> decorations) {
        super(ServerOpcodes.S_PROPERTY_PACK);
        put("mapId", mapId);

        ArrayNode arr = newArray();
        for (DecorationInstance d : decorations) {
            ObjectNode node = newObject();
            node.put("objId",      d._objId);
            node.put("name",       d._viewNote != null ? d._viewNote : "");
            node.put("x",          d._x);
            node.put("y",          d._y);
            node.put("heading",    2);
            node.put("propertyId", d._propertyId);
            node.put("pngid",      d._pngId);
            node.put("blocking",   d._blocking);
            node.put("footprintW", d._footprintW);
            node.put("footprintH", d._footprintH);
            node.put("offsetX",    d._offsetX);
            node.put("offsetY",    d._offsetY);
            node.put("action",     false);
            node.put("actionType", 0);
            node.put("value",      0);
            arr.add(node);
        }
        putArray("properties", arr);
    }

    /** 建立玩家裝飾的推送封包（單筆＝新放置一件）。 */
    public static S_PropertyPack ofDecorations(int mapId, List<DecorationInstance> decorations) {
        return new S_PropertyPack(mapId, decorations);
    }

    /**
     * 建立指定地圖的 {@link S_PropertyPack} 封包。
     *
     * @param mapId 地圖編號
     */
    public static S_PropertyPack of(int mapId) {
        return new S_PropertyPack(mapId);
    }
}
